Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -2676,6 +2736,40 @@ def visit_ClassDefNode(self, node):
         self.needs_closure = True
         return node
 
+class _FindComprehensionEvalWithTempExprNode(TreeVisitor):
+    def __init__(self):
+        super(_FindComprehensionEvalWithTempExprNode, self).__init__()
+        self.found_above_lambdas = []
+        self.lambdas = []
+
+    def visit_ComprehensionEvalWithTempExprNode(self, node):
+        if node.lazy_temp not in self.found_above_lambdas:
+            self.found_above_lambdas.append(node.lazy_temp)
+        self._visitchildren(node, None)
+
+    def visit_LambdaNode(self, node):
+        self.lambdas.append(node)
+        return # don't go any deeper
+
+    def visit_Node(self, node):
+        self._visitchildren(node, None) # necessary ?
+
+    @property
+    def found(self):
+        # this feels like duplicating the tracking done in
+        # "comp_input_variables" but I can't find a better way of doing it
```

Comment:
Yes, it does. Can we maybe call `create_class_from_scope()' *after* `visitchildren`?
Repeated calls to `tree_contains()` aren't entirely free of cost, and a single-pass traversal is always preferable.

Also, I don't understand what `found_above_lambdas` means here? In what way "above"? What it this class trying to achieve?

Reply:
This was definitely a bit of code I'm not really happy with. I'm trying to find `Comp...ResultRef`s that are used both inside and outside of a generator expression (using the base `LambdaNode` to find it). So it's first searching for all the `Comp...ResultRef` between the current node and whatever generator expressions exist, then looking to see which ones also appear in the generator expression.

It does seem an inefficient way of doing it so I'll definitely try to replace this approach.